File: iosgl/iosgl/bt/core/Context3D.m
//
//  Context3D.m
//  iosgl
//
//  Created by zhao on 8/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//

#import "Context3D.h"
#import "Matrix3D.h"

@implementation Context3D
- (instancetype)init
{
    self = [super init];
    if (self) {
        [self setupContext];
    }
    return self;
}
-(void)setupContext
{
    
    self.gl = [[EAGLContext alloc]initWithAPI:kEAGLRenderingAPIOpenGLES3];
    
    if( self.gl==NULL)
    {
        NSLog(@"Create Context Failed!");
        return;
    }
    if(![EAGLContext setCurrentContext: self.gl])
    {
        NSLog(@"setCurrentContext failed!");
        return;
    }
    
    
}
-(void)setVc3fv:(Shader3D*)shader name:(GLchar*)name data:(GLfloat*)data;
{
    glUniform3fv( glGetUniformLocation( shader.program, name), 1, data);
}
-(void)setVcMatrix3fv:(Shader3D*)shader name:(GLchar*)name data:(GLfloat*)data;
{
    glUniformMatrix3fv(  glGetUniformLocation( shader.program,name), 1, GL_TRUE,  data);
}
-(void)setVcMatrix4fv:(Shader3D*)shader name:(GLchar*)name data:(GLfloat*)data;
{
    glUniformMatrix4fv(  glGetUniformLocation( shader.program,name), 1, GL_TRUE,  data);
}
-(void)pushVa:(GLuint)dataBuffer;
{
     glBindBuffer(GL_ARRAY_BUFFER, dataBuffer);
}
-(void)setVaOffset:(Shader3D*)shader name:(GLchar*)name dataWidth:(int)dataWidth stride:(int)stride offset:(int)offset;
{
    GLuint position = glGetAttribLocation( shader.program,name);
    glEnableVertexAttribArray(position);
    glVertexAttribPointer(position, dataWidth, GL_FLOAT, GL_FALSE,stride, (GLfloat *)NULL+offset);
}
-(void)drawCall:(GLuint)ibuff numTril:(int)numTril;
{
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, ibuff);
    glDrawElements(GL_TRIANGLES, numTril, GL_UNSIGNED_INT, 0);
}
-(void)setRenderTexture:(GLKTextureInfo*)texture;
{
    glBindTexture(  texture.target,  texture.name);
}
@end

File: androld/app/src/main/java/z3d/core/Context3D.java
package z3d.core;

import android.opengl.GLES20;

import java.nio.FloatBuffer;
import java.nio.ShortBuffer;

import z3d.program.Shader3D;

public class Context3D {

    public void pushVa(FloatBuffer buf)
    {

    }
    public void  setProgame(int program)
    {
        GLES20.glUseProgram(program);
    }
    public void setVa(int dataId,int dataWidth,FloatBuffer dataBuffer)
    {
        GLES20.glVertexAttribPointer(dataId,dataWidth, GLES20.GL_FLOAT, false, dataWidth*4, dataBuffer);


    }

    public void setVaOffset(Shader3D shader3d, String name)
    {
        GLES20.glEnableVertexAttribArray(GLES20.glGetAttribLocation(shader3d.program, name));
    }
    public void  setVcMatrix4fv(Shader3D shader3d, String name , float[] m)
    {
        GLES20.glUniformMatrix4fv(GLES20.glGetUniformLocation(shader3d.program,name),1,false,m,0);
    }
    public void drawCall(ShortBuffer dataBuffer, int num)
    {

        GLES20.glDrawElements(GLES20.GL_TRIANGLES,num, GLES20.GL_UNSIGNED_SHORT,dataBuffer);


    }
}